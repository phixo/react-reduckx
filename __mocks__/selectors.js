import { createSelector } from 'reselect';

const branchSelector = state => state.mock;

const mockDataSelector = createSelector(
    branchSelector,
    state => state && state.data
);

const mockCounterSelector = createSelector(
    branchSelector,
    state => state && state.counter
);

export { branchSelector, mockCounterSelector, mockDataSelector };
