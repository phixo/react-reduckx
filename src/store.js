import React, {
    createContext,
    useContext,
    useEffect,
    useReducer,
    useState,
} from 'react';
import PropTypes from 'prop-types';

const createStore = (reducer, initialState) => {
    const ReduxContext = createContext();

    const ReduxProvider = ({ children }) => {
        const [isInitialized, setInitialized] = useState(false);
        const [state, dispatch] = useReducer(reducer, initialState);

        useEffect(() => {
            dispatch({ type: '@@REDUCKX/INIT' });
            setInitialized(true);
        }, []);

        return (
            <ReduxContext.Provider value={{ state, dispatch }}>
                {isInitialized && children}
            </ReduxContext.Provider>
        );
    };

    ReduxProvider.defaultProps = {
        children: null,
    };

    ReduxProvider.propTypes = {
        children: PropTypes.node,
    };

    const useRedux = (mapState, actionsMap, context = null) => {
        const { state, dispatch } = useContext(context || ReduxContext);
        const mappedState = mapState(state);

        const mappedActions = Object.keys(actionsMap).reduce((acc, key) => {
            const actionCreator = actionsMap[key];

            const fn = (...args) => {
                const action = actionCreator(...args);

                if (typeof action !== 'function') {
                    return dispatch(action);
                }
                return action({ getState: () => state, dispatch });
            };
            return { ...acc, [key]: fn };
        }, {});

        return [mappedState, mappedActions, dispatch];
    };

    return { ReduxProvider, useRedux };
};

export default createStore;
