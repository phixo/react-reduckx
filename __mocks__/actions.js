import axios from 'axios';
import { createAsyncAction } from '../src';
import { ACTION_ASYNC, ACTION_SYNC } from './action-types';

const mockAsyncActionSuccess = createAsyncAction(ACTION_ASYNC, id =>
    axios.get(`/mock/${id}`)
);

const mockAsyncActionFail = createAsyncAction(ACTION_ASYNC, () =>
    axios.get(`/mock/fail`)
);

const mockSyncAction = payload => ({
    type: ACTION_SYNC,
    payload,
});

export { mockAsyncActionSuccess, mockAsyncActionFail, mockSyncAction };
