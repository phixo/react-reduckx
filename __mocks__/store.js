import { combineReducers, createStore } from '../src';
import mockReducer, { initialState as mockState } from './reducers';

const rootReducer = combineReducers({
    mock: mockReducer,
    foo: mockReducer,
});

const initialState = {
    mock: { ...mockState },
};

export const { ReduxProvider, useRedux } = createStore(
    rootReducer,
    initialState
);
