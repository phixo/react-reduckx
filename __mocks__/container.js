import React, { useEffect } from 'react';
import { useRedux } from './store';

import { mapState, actionsMap } from './defaults';

const MockContainer = () => {
    const [{ counter, data }, { syncFn, asyncFn }] = useRedux(
        mapState,
        actionsMap
    );

    useEffect(() => {
        try {
            syncFn(1);
            asyncFn(1);
        } catch (error) {
            throw Error(error);
        }
    }, []);

    return (
        <>
            {counter} {JSON.stringify(data)}
        </>
    );
};

export default MockContainer;
