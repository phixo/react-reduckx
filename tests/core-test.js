import React, { createContext } from 'react';
import { act } from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import {
    renderHook,
    cleanup,
    act as actHook,
} from 'react-hooks-testing-library';

import mockAxios from '../__mocks__/axios';
import { actionsMap, mapState } from '../__mocks__/defaults';
import MockContainer from '../__mocks__/container';
import { ReduxProvider, useRedux } from '../__mocks__/store';

const mockReduckx = {
    dispatch: jest.fn(),
    state: jest.fn(() => {}),
};

describe('Store', () => {
    afterEach(() => {
        mockAxios.reset();
    });

    describe('ReduxProvider', () => {
        it('Should render properly', () => {
            const el = (
                <ReduxProvider>
                    <MockContainer />
                </ReduxProvider>
            );
            const tree = TestRenderer.create(el);

            act(() => {
                tree.update(el);
            });

            expect(tree).toMatchSnapshot();
        });
    });

    describe('useRedux', () => {
        const dispatchSpy = jest.spyOn(mockReduckx, 'dispatch');

        afterEach(() => {
            cleanup();
            dispatchSpy.mockClear();
        });

        it('Should return props with useRedux hook', () => {
            const context = createContext(mockReduckx);
            const { result } = renderHook(() =>
                useRedux(mapState, actionsMap, context)
            );
            const [, { syncFn, asyncFn }] = result.current;

            actHook(() => {
                syncFn(10);
            });

            expect(dispatchSpy).toBeCalledWith({
                payload: 10,
                type: 'MOCK/ACTION_SYNC',
            });

            actHook(() => {
                dispatchSpy.mockClear();
                asyncFn(100);
            });

            expect(dispatchSpy).toBeCalledWith({
                options: 100,
                type: 'MOCK/ACTION_ASYNC_PENDING',
            });
        });
    });
});
