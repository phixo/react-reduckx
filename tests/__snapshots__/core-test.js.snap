// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store ReduxProvider Should render properly 1`] = `
Array [
  "1",
  " ",
  "{}",
]
`;
